ansibledriver/service/ansible.py

```diff
@@ -221,127 +228,280 @@ def __init__(self, ansible_properties, request_id, lifecycle, display=None):
+    def v2_playbook_on_vars_prompt(self, varname, private=True, prompt=None, encrypt=None, confirm=False, salt_size=None, salt=None, default=None, unsafe=None):
+        """
+        Called when a var_prompt is used in a playbook, which we can't support because the playbook is not running in an interactive shell
+        """
+        logger.debug('v2_playbook_on_vars_prompt: {0}'.format(varname))
+        if self.ansible_properties.log_progress_events:
+            event = VarPromptEvent(var_name=varname)
+            self.event_logger.add(event)
 
     def runner_on_no_hosts(self):
         logger.debug('runner_on_no_hosts')
 
     def v2_runner_retry(self, result):
-        logger.debug('v2_runner_retry {0}'.format(result))
+        """
+        Called when a task is retried
+        """
+        logger.debug('v2_runner_retry: {0}'.format(result))
+        if self.ansible_properties.log_progress_events:
+            host_name = result._host.get_name().strip()
+            delegated_vars = result._result.get('_ansible_delegated_vars', None)
+            if delegated_vars is not None:
+                delegated_host_name = delegated_vars['ansible_host']
+            else:
+                delegated_host_name = None
+            self._clean_results(result._result, result._task.action)
+            task_name = result._task.get_name().strip()
+            event = TaskRetryOnHostEvent(task_name, host_name, result._result, delegated_host_name=delegated_host_name)
+            self.event_logger.add(event)
 
     def v2_runner_on_start(self, host, task):
-        logger.debug('v2_runner_on_start {0} {1}'.format(host, task))
+        """
+        Called when a task starts on a particular host (Ansible v2.8+)
+        """
+        logger.debug('v2_runner_on_start: host={0}, task={1}'.format(host, task))
+        if self.ansible_properties.log_progress_events:
+            task_name = task.get_name().strip()
+            host_name = host.get_name().strip()
+            event = TaskStartedOnHostEvent(task_name=task_name, host_name=host_name)
+            if not task.no_log:
+              # Include args if the task has not been configured with the no_log option
+              for arg_name, arg_value in task.args.items():
+                event.args[str(arg_name)] = str(arg_value)
+            else:
+              event.args_hidden = True
+            self.event_logger.add(event)
 
     def runner_on_failed(self, host, res, ignore_errors=False):
-        logger.debug('runner_on_failed {0} {1}'.format(host, res))
+        logger.debug('runner_on_failed: host={0}, result={1}'.format(host, res))
+    
+    def _log_event_for_failed_task(self, result, is_item=False):
+        if self.ansible_properties.log_progress_events:
+            host_name = result._host.get_name().strip()
+            delegated_vars = result._result.get('_ansible_delegated_vars', None)
+            if delegated_vars is not None:
```

Comment:
Potential NPE on result.* properties?

Reply:
Again, looking at https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/callback/default.py, my understanding is that result is always set, as is result._host and result._result.